src/main/java/com/mraof/minestuck/entity/item/ReturningProjectileEntity.java

```diff
@@ -0,0 +1,105 @@
+package com.mraof.minestuck.entity.item;
+
+import com.mraof.minestuck.client.renderer.entity.RendersAsItem;
+import com.mraof.minestuck.item.MSItems;
+import net.minecraft.entity.*;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
+import net.minecraft.entity.projectile.ProjectileItemEntity;
+import net.minecraft.item.Item;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.network.IPacket;
+import net.minecraft.util.DamageSource;
+import net.minecraft.util.SoundCategory;
+import net.minecraft.util.SoundEvents;
+import net.minecraft.util.math.*;
+import net.minecraft.world.World;
+import net.minecraftforge.fml.network.NetworkHooks;
+
+public class ReturningProjectileEntity extends ProjectileItemEntity implements RendersAsItem
+{
+	private int cooldownTicks;
+	private int bounce;
+	
+	public ReturningProjectileEntity(EntityType<? extends ReturningProjectileEntity> type, World worldIn)
+	{
+		super(type, worldIn);
+	}
+	
+	public ReturningProjectileEntity(EntityType<? extends ReturningProjectileEntity> type, double x, double y, double z, World worldIn)
+	{
+		super(type, x, y, z, worldIn);
+	}
+	
+	public ReturningProjectileEntity(EntityType<? extends ReturningProjectileEntity> type, LivingEntity livingEntityIn, World worldIn)
+	{
+		super(type, livingEntityIn, worldIn);
+	}
+	
+	@Override
+	protected void onImpact(RayTraceResult result)
+	{
+		PlayerEntity throwerPlayer = (PlayerEntity) this.getThrower();
+		if(throwerPlayer != null){
```

Comment:
Again, why a player thrower requirement for all of the below?

Reply:
The game crashes on impact without it, or at least thats what I remember happens

Reply:
It appear to only be used as a nonnull player for the cooldownTicks part (and resetThrower(), but it gets and checks the player on its own).
Ideally I'd think that the entity should still work in most manners, even with a null thrower or a non-player thrower. But what are your thoughts on this?
Independently of if you let more stuff happen with a non-player thrower or not, you should probably check that the type of the thrower is an instance of EntityPlayer before casting it to one.

Reply:
Just now I tried to make it work independently of a player but I could not get it to work fully. It would make sense to have it work with non-players, but because many of the properties are dependent on the player and the specific item, I dont seem to have the coding prowess necessary to solve the issue. It failed to render among other things

Reply:
With help, player dependency has been removed!